src/puppetlabs/puppetdb/client.clj

```diff
@@ -28,21 +28,21 @@
    submission. Alternately accepts a command-map object (such as those
    returned by `parse-command`). Returns the server response."
   ([base-url
-    certname :- s/Str
+    certname :- (s/maybe s/Str)
     command :- s/Str
     version :- s/Int
     payload]
    (submit-command-via-http! base-url certname command version payload nil))
   ([base-url
-    certname :- s/Str
+    certname :- (s/maybe s/Str)
     command :- s/Str
     version :- s/Int
     payload :- {s/Any s/Any}
     timeout]
    (let [body (json/generate-string payload)
          url-params (str
-                     (format "?command=%s&version=%s&certname=%s"
-                             (str/replace command #" " "_") version certname)
+                     (format (str "?command=%s&version=%s" (when certname "&certname=%s"))
+                             (str/replace command #" " "_") version (when certname certname))
```

Comment:
I think we may be able to cut down the moving pieces w.r.t handling the number of arguments to `format` if we move certname to its own format and let the outside-most `str` call stringify whatever is there, similarly to `producer_timestamp` below.

```
(str
  (format (str "?command=%s&version=%s") (str/replace command #" " "_") version)
  (when certname
    (format "&certname=%s" certname))
  (when-let [producer_timestamp (-> payload :producer_timestamp str)]
    (format "&producer-timestamp=%s" producer_timestamp)))
```

Reply:
yep, that's cleaner for sure. updated